fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,482 @@
+                    if (accept(oldState)) {
+                        result.push_back(Token(stateToType(oldState), munchedInput));
+                        munchedInput = "";
+                        state = start();
+                    } else {
+                        if (failed(state)) {
+                            munchedInput += *inputPosn;
+                        }
+                        throw TokenFailure("ERROR: Simplified maximal munch failed on input: " + munchedInput);
+                    }
+                }
+            }
+            return result;
+        }
+};
+
+std::vector<Token> scan(const std::string &input) {
+    static ASTstates rules;
+    std::vector<Token> tokens = rules.simplifiedMaximalMunch(input);
+    std::vector<Token> newTokens;
+    for (int i = 0; i < tokens.size(); i++) {
+        if (tokens[i].getType() != Token::Type::COMMENT && tokens[i].getType() != Token::Type::WHITESPACE) {
+			newTokens.push_back(tokens[i]);
+        }
+    }
+    return newTokens;
+}
+
+
+std::ostream &operator<<(std::ostream &out, const Token &tok) {
+    switch(tok.getType()) {
+        case Token::ID: out << "ID"; break;
+        case Token::STRING_: out << "STRING_"; break;
+        case Token::INT: out << "INT"; break;
+        case Token::DOUBLE_: out << "DOUBLE_"; break;
+        case Token::AT:  out << "AT"; break;
+        case Token::ZERO:  out << "ZERO"; break;
+        case Token::QUOT:  out << "QUOT"; break;
+        case Token::LPAREN:  out << "LPAREN"; break;
+        case Token::RPAREN:  out << "RPAREN"; break;
+        case Token::LPAREN_SQ:  out << "LPAREN_SQ"; break;
+        case Token::RPAREN_SQ:  out << "RPAREN_SQ"; break;
+        case Token::BECOMES: out << "BECOMES"; break;
+        case Token::COMMA: out << "COMMA"; break;
+        case Token::HEXINT:  out << "HEXINT"; break;
+        default: throw TokenFailure("ERROR: Failed to determine input: " + tok.getValue());
+    }
+    out << " " << tok.getValue();
+    return out;
+}
+
+class TokenIter {
+    private:
+        std::vector <Token>::iterator it;
+        std::vector <Token>::iterator end;
+    public:
+        TokenIter(std::vector <Token>::iterator it, std::vector <Token>::iterator end)
+            : it(it), end(end) {}
+        std::vector <Token>::iterator operator++() { return it++;}
+        bool is_end() { return it == end;}
+        std::vector <Token>::iterator peek() {
+            std::vector <Token>::iterator temp = it;
+            temp++;
+            return temp;
+        }
+        std::vector <Token>::iterator getIt() { return it; }
+        std::vector <Token>::iterator getEnd() { return end; }
+};
+
+ASTValue* buildNodeValue(Token token) {
+    Token::Type tokenType = token.getType();
+    std::string tokenValue = token.getValue();
+    if (tokenValue.size() > 1 && tokenType == Token::STRING_) { 
+        tokenValue = tokenValue.substr(1, tokenValue.size() - 2);
+    }
+    char * cstr = new char[tokenValue.size() + 1];
```

Comment:
More of a question for @Leonardo2718. How manages this memory? Do we ever free it?

Reply:
Currently, that memory is never freed. It's not ideal but that's how the yacc-generated parser worked. In fact, none of the AST nodes are ever freed because yacc does not make it easy to track ownership. To avoid having to rewrite a bunch of other code, I suggested continuing to leak the memory. Eventually, the proper fix will be to refactor the AST so it properly manages memory, which will be another significant undertaking.

Reply:
Ok, I agree. We should tackle this when it becomes an issue. I imagine once the number of tests we execute increases we may start to run out of memory as each test case would leak memory and there are currently 10s of thousands of tests that run.